test/corpus/expressions.txt

```diff
@@ -4149,3 +4149,231 @@ asm ("nop")
     text: (string))
   (asm
     text: (string)))
+
+================
+as pseudo method
+================
+
+node.as(LibXML::Node*)
+as(Int32 | LibXML::Node*)
```

Comment:
This should not be valid.

Reply:
How so? The latter is calling `as` on self

Reply:
```crystal
icr:1> class MyClass
icr:2>   def hello
icr:3>     as(String)
icr:4>   end
icr:5> end
 => nil
icr:6> a = MyClass.new
 => #<MyClass:0x11b0b1b90>
icr:7> a.hello
error in line 1
Error: instantiating 'MyClass#hello()'


error in line 3
Error: can't cast MyClass to String
```

Reply:
These pseudo-methods are heavily contextual so it doesn't make sense to accept them as valid in that sense. I don't suppose we could represent them as erroneous in this context either, but maybe as a different node?

Reply:
This is valid code, shouldn't tree-sitter parse it?
``` crystal
class Foo
  def self_or_pointer
    as(self | UInt8*)
  end
end
```

Reply:
If I understand what you're referring to, the issue of `as` only working in specific contexts is a semantic error, not a syntactic one. Should be handled by semantic analysis done afterwards on the tree.